typst: add initial support for typst packages (#369283)
NixOS/nixpkgs: doc/languages-frameworks/typst.section.md

> # Typst {#typst}

Reviewer: I modified a bit the documentation with better clarity, grammar, and structure while taking care of preserving the technical accuracy.

```
# Typst {#typst}

Typst can be configured to include packages from [Typst Universe](https://typst.app/universe/) or custom packages.

## Custom Environment {#typst-custom-environment}

You can create a custom Typst environment with a selected set of packages from **Typst Universe** using the following code. It is also possible to specify a specific version of a Typst package (e.g., `cetz_0_3_0`). A package without a version number will always refer to its latest version.

```nix
typst.withPackages (p: with p; [
  polylux
  cetz_0_3_0
])
```

### Handling Outdated Package Hashes

Since **Typst Universe** does not provide a way to fetch a package with a specific hash, the package hashes in `nixpkgs` can sometimes be outdated. To resolve this issue, you can manually override the package source using the following approach:

```nix
typst.withPackages.override (old: {
  typstPackages = old.typstPackages // {
    polylux = old.typstPackages.polylux.overrideAttrs (oldPolylux: {
      src = fetchurl {
        url = oldPolylux.src.url;
        hash = yourNewPolyluxHash;
      };
    });
  };
}) (p: with p; [
  polylux
  cetz_0_3_0
])
```

## Custom Packages {#typst-custom-packages}

`Nixpkgs` provides a helper function, `buildTypstPackage`, to build custom Typst packages that can be used within the Typst environment. However, all dependencies of the custom package must be explicitly specified in `typstDeps`.

Here's how to define a custom Typst package:

```nix
{ buildTypstPackage, typstPackages, fetchurl }:

buildTypstPackage {
  pname = "my-typst-package";
  version = "0.0.1";
  src = fetchurl { ... };
  typstDeps = with typstPackages; [ cetz_0_3_0 ];
}
```

### Package Scope and Usage

By default, every custom package is scoped under `@preview`, as shown below:

```typst
#import "@preview/my-typst-package:0.0.1": *
```

Since `@preview` is intended for temporary or experimental modifications, it is recommended to use this approach **only for patching existing packages** from **Typst Universe**.

On the other hand, **local packages** are scoped under `@local`, but they are **not** considered part of the Typst environment. This means that local packages must be manually linked to the Typst compiler if needed.
```

Typst can be configured to include packages from Typst Universe or customized packages.

## Custom Environment {#typst-custom-environment}

Creating a custom Typst environment with a selected set of packages from Typst Universe can be done with the following code. It is possible to specify a specific version of a Typst package (e.g., `cetz_0_3_0`). A package without a version number refers to its latest version.

```nix
typst.withPackages (p: with p; [
  polylux
  cetz_0_3_0
])
```

> Since Typst Universe doesn't provide an way to fetch a package with a specific hash, it can be common that Typst packages' hash in nixpkgs are outdated. The example below can fix this issue.

Reviewer: ```suggestion
Since Typst Universe doesn't provide a way to fetch a package with a specific hash, it can be common that Typst packages' hash in nixpkgs are outdated. The example below can fix this issue.
```

```nix

> typst.withPackages.override (old: {

Reviewer: Wouldn't it be easier to use overrideAtrs ?

Reviewer 2: 1. `typst.withPackages` is a function (callable), which can only be overridden with `<function>.override`.
1. Even with `overrideAttrs`, `override` is responsible for package dependency overriding, which is relevant here.

  typstPackages = old.typstPackages // {
    polylux = old.typstPackages.polylux.overrideAttrs (oldPolylux: {

>       src = fetchurl {

Reviewer: I was suggesting that in a previous comment too.

        url = oldPolylux.src.url;
        hash = yourNewPolyluxHash;
      };
    });
  };
}) (p: with p; [
  polylux
  cetz_0_3_0
])
```

## Custom Packages {#typst-custom-packages}

Nixpkgs provide a helper `buildTypstPackage` to build custom Typst packages understood by the Typst environment. Any dependency of the custom package is required to be manually specified in `buildTypstpackage`. Here is how to do it.

```nix
{ buildTypstPackage, typstPackages, fetchurl }:

buildTypstPackage {
  pname = "my-typst-package";
  version = "0.0.1";
  src = fetchurl { ... };
  typstDeps = with typstPackages; [ cetz_0_3_0 ];
}
```

> Note that every custom package is scoped in `@preview`, as shown below. To not violating the semantics of `@preview`, it is recommended to use this only for patching existing packages from Typst Universe.

Reviewer 3: ```suggestion
Note that every custom package is scoped in `@preview`, as shown below. In order to avoid violating the semantics of `@preview`, it is recommended to use this only for patching existing packages from Typst Universe.
```

I am not sure if "To not violating" is grammatically correct.

# Typst {#typst}

Typst can be configured to include packages from [Typst Universe](https://typst.app/universe/) or custom packages.

## Custom Environment {#typst-custom-environment}

You can create a custom Typst environment with a selected set of packages from **Typst Universe** using the following code. It is also possible to specify a Typst package with a specific version (e.g., `cetz_0_3_0`). A package without a version number will always refer to its latest version.

```nix
typst.withPackages (p: with p; [
  polylux
  cetz_0_3_0
])
```

### Handling Outdated Package Hashes {#typst-handling-outdated-package-hashes}

Since **Typst Universe** does not provide a way to fetch a package with a specific hash, the package hashes in `nixpkgs` can sometimes be outdated. To resolve this issue, you can manually override the package source using the following approach:

```nix
typst.withPackages.override (old: {

>   typstPackages = old.typstPackages // {
>     polylux = old.typstPackages.polylux.overrideAttrs (oldPolylux: {
>       src = oldPolylux.src.overrideAttrs {
>         outputHash = YourUpToDatePolyluxHash;
>       };
>     });
>   };

Reviewer 2: ```suggestion
  typstPackages = old.typstPackages.extend (final: previous: {
    polylux_0_4_0 = previous.polylux_0_4_0.overrideAttrs (oldPolylux: {
      src = oldPolylux.src.overrideAttrs {
        outputHash = YourUpToDatePolyluxHash;
      };
    });
  });
```

}) (p: with p; [

>   polylux

Reviewer 2: ```suggestion
  polylux_0_4_0
```
